### PR TITLE
[BEAM-6324] Add filter to query for CassandraIO.Read

### PR DESCRIPTION
**Pull request by srfrnk**
CassandraIO.Read doesn't support using WHERE to filter the input at the source (In Cassandra) which might provide great performance boost.

This PR adds the ability to provide a filter to the query.
------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by srfrnk** [comment]
Implements https://issues.apache.org/jira/browse/BEAM-6324

**Comment by timrobertson100** [comment]
Thank you for this @srfrnk and please accept our apologies this went unnoticed. In the future please don't be shy about sending a reminder to the dev@ list to prompt action.

Can you please explore what is feasible in adding a test for this? That might be something that verifies the predicate is pushed down to Cassandra or if that proves overly difficult, perhaps simplify verifying using a [logger approach](https://github.com/apache/beam/blob/master/sdks/java/io/solr/src/test/java/org/apache/beam/sdk/io/solr/SolrIOTest.java#L81).

**Comment by timrobertson100** [comment]
Retest this please.

File: sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java [hunk 1]
```diff
@@ -237,37 +237,44 @@ static long getEstimatedSizeBytes(List<TokenRange> tokenRanges) {
     for (List<RingRange> split : splits) {
       List<String> queries = new ArrayList<>();
       for (RingRange range : split) {
-        Select.Where builder = QueryBuilder.select().from(spec.keyspace(), spec.table()).where();
         if (range.isWrapping()) {
-          // A wrapping range is one that overlaps from the end of the partitioner range and its
+          // A wrapping range is one that overlaps from the end of the partitioner range
+          // and its
           // start (ie : when the start token of the split is greater than the end token)
```
Comment (timrobertson100) [inline sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java]: Nit: please correct line breaks

**Comment by echauchot** [comment]
thanks @srfrnk thanks for your work! indeed this PR got unseen.

**Comment by echauchot** [comment]
As a general comment, I think it is time to replace the mock in the CassandraIO Utest by a proper embedded cassandra instance. See [this](https://lists.apache.org/thread.html/960335730dcf93e8bc82a1df0caceb5f29d992f25a0f1bc8c2eb760f@%3Cdev.beam.apache.org%3E) discussion on the ML. I have already used [this embedded server](https://github.com/jsevellec/cassandra-unit ) in the past and it works well

**Comment by echauchot** [comment]
Another general comment (I leave detailed review to Tim as he already started) : IIRC you cannot filter with WHERE clause on Cassandra unless the column on which you filter is either a clustering column or a primary column. So you need to check this before doing a push down predicate.

**Comment by srfrnk** [comment]
@timrobertson100 Thanks. Will make the required fixes + take a look at adding a test. Hopefully next week.
@echauchot Thanks. I am aware that Cassandra hold many constraints over what you can and cannot do in a query (and specifically in the Where clause) - however when using a String Where clause I don't think any such tests can be done very safely. I suppose any incorrect Where cluase would come back as an error and shown to the user. I will need to test that though to confirm.

In general this is just a workaround to be able to use Cassandra with a reasonable amount of  performance. The real solution has to use the QueryBuilder. Alas since Clause (and other classes by Datastax) is not Serializable this is impossible to do. I have tried to [PR for a change to allow this but it was rejected](https://github.com/datastax/java-driver/pull/1163). Currently I'm using [my own forked repo of the datastax java-driver ](https://github.com/srfrnk/java-driver) which has the compiled JAR pushed so I can use it from other projects directly. Until such version is introduces I don't think better code can be done in here. WDYT?

**Comment by echauchot** [comment]
I'm surprised that a withQuery() does not already exist on CassandraIO. I guess it has to do with dataset size estimation for splitting (if you don't know the size you cannot split). And I agree that avoiding full table scan is very important.  That being said, if you add a `withWhereClause()`  method why not add the complete withQuery() ? We then need to solve this problem: how can we get  the size of the input dataset without reading it all in advance? (take a look at JDBCIO, I don't remember how it was solved) Once the size is known, then splitting is done with the range query. Regarding using QueryBuilder or String query, the former adds more security but we can leave the responsibility to the user to write correct queries as it is done in other IOs, so I would vote for the latter because it works out of the box without having serialization issues.

Regarding use of non primary/cluster columns, we can accept raising the error to the user but it needs to be tested that it actually raises an error in that case.

**Comment by chamikaramj** [comment]
Thanks for the contribution.

I don't have any additional comments so will defer to other reviewers.

**Comment by echauchot** [comment]
@srfrnk regarding my comment on using a embedded cassandra server rather than the mock. Actually I had already opened a ticket for that a long time ago (https://issues.apache.org/jira/browse/BEAM-4164?jql=project%20%3D%20BEAM%20AND%20reporter%20in%20(echauchot)%20AND%20text%20~%20%22cassandra%22). It is scope creep toward this PR. So I'll deal with it

File: sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java (hunk 1 above)
Comment (srfrnk) [reply sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java]: done.. please recheck

**Comment by srfrnk** [comment]
@timrobertson100 , @echauchot I think the best place to add a test for this is `./sdks/java/io/cassandra/src/test/java/org/apache/beam/sdk/io/cassandra/CassandraIOIT.java`

However it seems like the integration tests are no longer being run.
`gradle build` certainly doesn't and the comment at the head of the file only mentions how to run them using maven.

Can you please tell me if I can use these at all? If so how? thx.

**Comment by srfrnk** [comment]
@echauchot Re. size estimation - I can't see `./sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java` doing any such thing. Can you maybe point the line numbers out?

**Comment by timrobertson100** [comment]
> Can you please tell me if I can use these at all? If so how? thx.

You should be able to run them in gradle similar to this approach:
https://github.com/apache/beam/blob/master/sdks/java/io/elasticsearch-tests/elasticsearch-tests-5/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOIT.java#L41

Assuming it works, can you please update the JDoc as well? I don't know if you are aware but a few months ago Beam moved to gradle from maven and we're still cleaning up a few things like that documentation as it is found.

Thank you.

**Comment by srfrnk** [comment]
@timrobertson100 , @echauchot  Test was added. Please review.

**Comment by srfrnk** [comment]
> Run Java PreCommit

The build logs show:
13:59:27 > Task :beam-runners-java-fn-execution:test
13:59:27 
13:59:27 org.apache.beam.runners.fnexecution.data.GrpcDataServiceTest > testMessageReceivedBySingleClientWhenThereAreMultipleClients FAILED
13:59:27     java.lang.AssertionError at GrpcDataServiceTest.java:105

Seems unrelated to any of my changes. WDYT?

**Comment by timrobertson100** [comment]
FYI: You can trigger a build by entering a comment like "Run Java PreComit", "Run Java..." here. You don't need to commit a "touch". The failure is a flakey test unrelated to your commits as far as I can see.

**Comment by srfrnk** [comment]
> FYI: You can trigger a build by entering a comment like "Run Java PreComit", "Run Java..." here. You don't need to commit a "touch". The failure is a flakey test unrelated to your commits as far as I can see.

Thx :) Didn't know that. Your comment didn't seem to run it I think. Anyway the 'touch' commits seem to have run the build again and it looks like the code is good but maybe the tests aren't deterministic.

Anyway it looks like all tests are passing now.

Did you perhaps look at the code? Does it look good?

**Comment by echauchot** [comment]
Regarding using the ITests to test cassandra, I think it would be better to do it in UTests. I understand why you chose ITests, because you wanted a proper cassandra instance to test against. This is why I pointed out using an embedded cassandra instance. I'm precisely refactoring CassandraIO to include such an embedded instance in UTests and also to remove CassandraService which was mainly there to enable injecting FakeCassandraService mock.

**Comment by echauchot** [comment]
@srfrnk ok to make your tests in ITests. I will move them to UTests later on when I'm ready. I did not take a precise look at the code, I just wanted to discuss general directions. I'm ok to proceed with this PR as long as @timrobertson100 is ok. I'll let Tim finish his review. 

Regarding rewriting the Reader it is out of sope of this PR but definitely promising ! I see that you replace the source with a Pardo, I think that it deserves a general project discussion on whether we generalize this approach.

**Comment by srfrnk** [comment]
Thanks.
Re. the Pardo approach - I found this saying that ["ParDo is the recommended option"](https://beam.apache.org/documentation/io/developing-io-overview/).
Also I think "Splittable DoFn" would be the thing to replace both unbounded and bounded sources.
So that's why I was thinking of going that way. If there is such a discussion I'd be happy to see it.

**Comment by timrobertson100** [comment]
LGTM.

Thank you for the changes @srfrnk

**Comment by timrobertson100** [comment]
Can you please squash your commits into a single one and I will merge it @srfrnk 

If you prefer I do it, I will do so on Wednesday (sorry, travelling and with little time right now)

**Comment by srfrnk** [comment]
I've tried to squash them but it got wierd. I suspect because I merged from main repo between commits it may not be possible and since I'm new to squashing I prefer not to create any damage.
Hopefully you can manage that when possible.
Thanks.

**Comment by timrobertson100** [comment]
@srfrnk 
Can you please grant me commit access to your repo? I need to do a force push after the rebase and fixup and have never done that on a PR branch directly in the master project. I feel safer doing that on your branch in your repo if you don't object.

**Comment by srfrnk** [comment]
@timrobertson100 You should be able to now. Please confirm.

**Comment by timrobertson100** [comment]
I haven’t forgotten this but am off with sick family. It needs a rebase to latest master, and then to fix the one conflict (accept the imports if I remember correctly). I will do it as soon as I can unless someone beats me to it.

**Comment by srfrnk** [comment]
Thanks @timrobertson100! No rush - be well!

**Comment by echauchot** [comment]
I'll try to get PR https://github.com/apache/beam/pull/7751 merged before so that you can use the embedded in your test

**Comment by timrobertson100** [comment]
Can you please take a look @srfrnk and just confirm I didn't miss anything please?

I am afraid I had to rebase against master fixup and then do a force commit. I think it reflects your original intentions but is now a single commit and the branch in line with the current beam/master.

In the future, it will be easiest if you periodically rebase against master avoiding merge commits (`git fetch upstream master`, `git checkout <your_branch>`, `git rebase origin/master`). Following this approach makes it easy for you to finish with a squash of the last N commits into a single one.

**Comment by srfrnk** [comment]
@timrobertson100 Thanks. Will test it on the weekend and let you know.

**Comment by srfrnk** [comment]
run java precommit

**Comment by echauchot** [comment]
@srfrnk @timrobertson100, hi guys, please wait a bit for my other [PR](https://github.com/apache/beam/pull/7751) to be merge (with 2 days I think). That would allow you to better test the filter/query in UTests.

**Comment by timrobertson100** [comment]
Thanks @echauchot - will do.

@srfrnk I can imagine this is frustrating for you. Are you happy to adjust the tests when PR #7751 is merged or would you like me or @echauchot to tackle that?

**Comment by srfrnk** [comment]
@echauchot - Sure no prob.
@timrobertson100 - frustration level == 0 :) Will do that.

**Comment by echauchot** [comment]
@srfrnk thanks for your understanding! I know, that this PR as taken too long to start the review (because no reviewed picked it up automatically) at least we did our best to avoid scope creeping :) and took care of the embedded cassandra problem :)
By the way, FYI the experience with this PR not been assigned a reviewer has helped the community improve by changing the template PR to remind contributors to chose a reviewer and help them in that action.

**Comment by srfrnk** [comment]
@echauchot I'm glad it helped.
I believe we all share the goal of making Beam better.
Please update me on when your PR is merged so that I can redo the tests.

**Comment by echauchot** [comment]
@srfrnk I just merged my [PR](https://github.com/apache/beam/pull/7751), please rebase, refactor (there is no more CassandraService and CassandraServiceImpl) and adapt your UTests to the new embedded Cassandra instance

**Comment by srfrnk** [comment]
Thanks @echauchot - I hope to have to time this weekend. Will update.

**Comment by srfrnk** [comment]
@echauchot - I rebased the code. Had to manually re-add the changes but I think it was worth it. Your PR makes the tests much better IMHO.

@timrobertson100 - I hope this time I got everything right - the code, single commit, docs and tests. Could you please review?

**Comment by echauchot** [comment]
@srfrnk Thanks for rebasing, and also for adding your test as UTest !
Also, I just overlooked the PR (leaving the review to Tim) 2 coments:
- would it be complicated to add a more general withQuery instead of only withWhere?
- if too complicated I would advice to update the withWhere javadoc with a link to cassandra doc to warn the user that he should query on a primary/clustering column and tell him about the exception he will get otherwise.

**Comment by srfrnk** [comment]
I considered using the term Query rather then Where - however I suspect it may confuse users.
We are just injecting a partial filter to the Where clause of the cassandra Query here.
It will not allow setting the full Query.
It even doesn't allow setting the full Where clause indeed.
That's why I finally decided to use Where rather then Query in the property name.
WDYT?

**Comment by echauchot** [comment]
@srfrnk why not just allow the user to pass the query as a String like it is done in `ElasticsearchIO.Read.withQuery()` for ex ?

**Comment by srfrnk** [comment]
@echauchot I would like that of course - however I don't know how that can be achieved at this time. I was hoping at the beginning to be able to create an API similar to BigQueryIO which I used a lot and liked very much. It also has a `.withQuery` function similar to `ElasticsearchIO`. However with the current method where a query is buing built for the splits in `CassandraIO` I can't think of a safe and comfortable way of allowing the user bring his own query while managing the query for the splitting.
I finally settled on having the user supply a filter in the form of a partial `Where` clause.
Can you think of a better way?

**Comment by srfrnk** [comment]
Run JavaPortabilityApi PreCommit

**Comment by timrobertson100** [comment]
@srfrnk @echauchot 
I am running a workshop in Australia for the next week and a half and won't have much time to Beam. The code appears similar to the previous commit which was reviewed and changes were all applied - thank you! If @echauchot is ok with the tests I recommend this gets merged.

**Comment by srfrnk** [comment]
Thanks @timrobertson100, except for the new Unit Tests the rest is indeed identical.
I would be glad to merge and get this over with myself :)

**Comment by echauchot** [comment]
@srfrnk, basically split query only adds `where token >= startToken AND token <= endToken` so we could have a String builder that concatenates these split filters with the query provided by the user. But this can be done in a future PR. So please just add the javadoc user warning (cf above) and I will merge the PR as Tim is unavailable and he is ok with the PR.

**Comment by srfrnk** [comment]
@echauchot Could you please review the `javadoc` comment I added?

**Comment by srfrnk** [comment]
@echauchot Thanks. Can you explain "checkstyle fix"?
Will squash once you approve everything.

**Comment by echauchot** [comment]
@srfrnk yes, the current build fails because of checkstyle violations: see [gradle build scan](https://scans.gradle.com/s/6ve7ezdffo6zk)

**Comment by srfrnk** [comment]
@echauchot - Rebased. Hopefully all passes this time.

**Comment by echauchot** [comment]
@srfrnk checkstyle now passes but cassandraIO javadoc is incorrect: see https://scans.gradle.com/s/y4f3d4jbyzdcw/console-log?task=:beam-sdks-java-io-cassandra:javadoc

**Comment by srfrnk** [comment]
@echauchot - Looks like it's passing now.

**Pull request merged**